### ISSUE TITLE
How to avoid upload speed degrading with latency?

### ISSUE THREAD

**Issue by ncw** [pkg/sftp#158]
We've been integrating sftp package into [rclone](https://github.com/ncw/rclone) and it is working very well - a very nicely designed package, thank you.

However a user has noticed a performance problem which I've confirmed - the more latency in the internet connection, the slower the transfers.

I've been using [netelem](https://wiki.linuxfoundation.org/networking/netem) to set delays on localhost and trying transfers to and from a local sftp server.

  * At 1ms delay I get 100 MB/s
  * At 10ms delay I get 73 MB/s
  * At 100ms delay I get 9 MB/s

rclone uses the `File.ReadFrom` method to upload data.

I tried increasing `MaxPacket` in the client using a delay of 100ms, but anything greater than the default 32k just seemed to make things slower.

Any ideas on how to speed things up much appreciated.

Cheers

Nick

FYI Here is how I tested with rclone to a local sftp server

```
$ sudo tc qdisc change dev lo root netem delay 1ms
$ rclone -v copy /home/files/ISOs/ubuntu-16.04.1-desktop-amd64.iso TestSftp:
2017/02/20 22:28:02 INFO  : 
Transferred:   1.409 GBytes (100.970 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:       14.2s
$ rclone delete TestSftp:ubuntu-16.04.1-desktop-amd64.iso

$ sudo tc qdisc change dev lo root netem delay 10ms
$ rclone -v copy /home/files/ISOs/ubuntu-16.04.1-desktop-amd64.iso TestSftp:
2017/02/20 22:28:38 INFO  : 
Transferred:   1.409 GBytes (73.208 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:       19.7s
$ rclone delete TestSftp:ubuntu-16.04.1-desktop-amd64.iso

$ sudo tc qdisc change dev lo root netem delay 100ms
$ rclone -v copy /home/files/ISOs/ubuntu-16.04.1-desktop-amd64.iso TestSftp:
2017/02/20 22:31:34 INFO  : 
Transferred:   1.409 GBytes (9.362 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:     2m34.1s
$ rclone delete TestSftp:ubuntu-16.04.1-desktop-amd64.iso
```

**Comment by kardianos** [comment]
@ncw I suspect without something like https://mosh.org/ , this is just a property of ssh itself. Have you tried these benchmarks with the raw ssh package transferring a raw byte stream?

**Comment by ncw** [comment]
@kardianos I think you may be right.  This is the long fat pipe TCP problem.

No delay

```
$ scp /home/files/ISOs/ubuntu-16.04.1-desktop-amd64.iso sftptest@localhost:
ubuntu-16.04.1-desktop-amd64.iso 100% 1443MB 111.0MB/s   00:13  
```

100ms delay

```
$ scp /home/files/ISOs/ubuntu-16.04.1-desktop-amd64.iso sftptest@localhost:
ubuntu-16.04.1-desktop-amd64.iso 100% 1443MB   9.6MB/s   02:31    
```

What I don't understand is the original user report which said FileZilla could transfer much faster.  I'll go back to the user and ask for more clarification as FileZilla wasn't any faster in my tests.

**Comment by kardianos** [comment]
Oh, filezilla will open N simultaneous connections to transmit  a file,
this circumvents per connection rate limiters and link latencies.

**Comment by justusiv** [comment]
I am the OP.
I am under the impression that Filezilla does not do Segmented Downloading. While both tools are doing multi threaded downloading at 4 a piece.

**Comment by eikenb** [comment]
Looked at the rclone repo and it seems this has been resolved.

The problem is documented in https://github.com/ncw/rclone/issues/1158 and https://github.com/ncw/rclone/issues/1561.

The way Filezilla, rclone and others deal with the latency issue is by opening multiple SSH connections and distributing the file transfer over those. This means the problem is not with the SFTP library as it doesn't manage the SSH connections. So it is by necessity up to the program using this library to do that.

If you're curious about how rclone implemented the multi-ssh transfer pool, here is the relevant commit.
https://github.com/ncw/rclone/commit/47eab397ba6e5dbf3d6de1f71265334393f279c7

**Comment by ncw** [comment]
@eikenb I don't think this issue is fixed by connection pooling alas.  There seems to be some sort of fundamental slowdown with high latency links and this library.

Here is some evidence!

I set up a local sftp server and ran

    sudo tc qdisc add dev lo root netem delay 50ms

Check that ping localhost gives the expected delay

```
$ ping localhost
PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=100 ms
64 bytes from localhost (127.0.0.1): icmp_seq=2 ttl=64 time=101 ms
^C
```

(Use this to remove the delay)

    sudo tc qdisc del dev lo root netem

To make a 100ms roundtrip delay.

I can get about 18 MB/s with the sftp binary

    sftp sftptest@localhost:1GB

Whereas if I try this with rclone I get about 1.5 MB/s

    rclone --stats 1s -v copy TestSftp:1GB /tmp

If I try them both with no latency then they are very similar 105MB/s rclone vs 128 MB/s sftp.

I tried a comparison with netcat also.  So running a netcat server in one window

    netcat -k -l localhost 9999 </dev/zero

I measured the speed with netcat in another

    netcat localhost 9999 | pv >/dev/null

Which came about at about 25 MB/s.

Using a simple [go netcat replacement](https://play.golang.org/p/cVPXSu3m23M) I wrote I get a very similar 26 MB/s

    ~/Go/netcat localhost 9999 | pv >/dev/null

So it doesn't appear to be a problem with the go TCP stack.  I did try increasing SO_RCV_BUF with no improvements.

So I think this is something to do with this library - exactly what I'm not sure.

**Comment by puellanivis** [comment]
So, the reason it is not a problem with the TCP stack alone is because if a link has very high latency, but a predictable transfer rate, many TCP stacks will preemptively ACK data despite not actually having receiving it yet. However, if the application layer on top of the TCP requires extensive handshaking in any way as well, then this steady transfer rate presumption is broken, and the TCP stack cannot preemptively ACK these transmissions.

[The author of HPN-SSH noted](https://stackoverflow.com/questions/8849240/why-when-i-transfer-a-file-through-sftp-it-takes-longer-than-ftp) that because SSHv2 uses a multiplexed channel, it has to implement a flow-control on top of TCP, which would readily explain why latency ends up making this issue.

So, the problem lies in an emergent interaction between the SFTP and TCP protocols, as neither can fully coordinate their flow controls to ensure steady long-and-fat-pipe transfers, and thus, the need for clients to use “work arounds” such as parallel connections.

**Comment by eikenb** [comment]
@puellanivis That stackoverflow response by the HPN-SSH author was a nice find and definitely sheds some light on the issue.

After reading though it I checked and the x/crypto/ssh package uses the same max buffer size (2MB). Unfortunately it is a private field which cannot be changed. We might need to see if we can get this converted into a modifiable field to address this any more as, according to that post, it is how you get any real gains.

**Comment by ncw** [comment]
@puellanivis very useful link thank you.

I tried patching `x/crypto/ssh` thus

```patch
diff --git a/vendor/golang.org/x/crypto/ssh/channel.go b/vendor/golang.org/x/crypto/ssh/channel.go
index c0834c00..7a39eac4 100644
--- a/vendor/golang.org/x/crypto/ssh/channel.go
+++ b/vendor/golang.org/x/crypto/ssh/channel.go
@@ -20,7 +20,7 @@ const (
        // the minimum.
        channelMaxPacket = 1 << 15
        // We follow OpenSSH here.
-       channelWindowSize = 64 * channelMaxPacket
+       channelWindowSize = 8 * 64 * channelMaxPacket
 )
```

To increase the window to 16MB.  I also set the Recv and Send buffers on the tcp sockets

```patch
diff --git a/backend/sftp/sftp.go b/backend/sftp/sftp.go
index 1fe1553f..03f1aa6c 100644
--- a/backend/sftp/sftp.go
+++ b/backend/sftp/sftp.go
@@ -154,6 +154,30 @@ func Dial(network, addr string, sshConfig *ssh.ClientConfig) (*ssh.Client, error
        if err != nil {
                return nil, err
        }
+
+       const bufferSize = 16 * 1024 * 1024
+       if do, ok := conn.(interface{ SetReadBuffer(bytes int) error }); ok {
+               fs.Logf(nil, "Setting read buffer size to %v", bufferSize)
+               err = do.SetReadBuffer(int(bufferSize))
+               if err != nil {
+                       return nil, errors.Wrapf(err, "couldn't set read buffer size %v", bufferSize)
+               }
+       }
+       if do, ok := conn.(interface{ SetWriteBuffer(bytes int) error }); ok {
+               fs.Logf(nil, "Setting write buffer size to %v", bufferSize)
+               err = do.SetWriteBuffer(int(bufferSize))
+               if err != nil {
+                       return nil, errors.Wrapf(err, "couldn't set write buffer size %v", bufferSize)
+               }
+       }
+
        c, chans, reqs, err := ssh.NewClientConn(conn, addr, sshConfig)
        if err != nil {
                return nil, err
```

However it made no difference whatsoever :-(

----

Taking a step back I adapted an ssh client I found to test out the ssh part of the connection ([playground](https://play.golang.org/p/OUsdo9s7NVa)):

I tested it like this...

```
$ ./ssh_client ncw localhost 'cat /dev/zero' | pv -
Running command: cat /dev/zero
^C64MiB 0:00:10 [19.9MiB/s] [                               <=>                                                                                                                              ]
```

vs openssh

```
$ ssh ncw@localhost 'cat /dev/zero' | pv -
^C.0MiB 0:00:06 [19.7MiB/s] [               <=>                                                                                                                                              ]
```

Over the 100ms roundtrip connection

```
$ ping localhost
PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=100 ms
64 bytes from localhost (127.0.0.1): icmp_seq=2 ttl=64 time=100 ms
^C
```

So the go ssh layer appears to perform exactly the same as the openssh layer and at the 19 MB/s rate I'm expecting.

So I think the problem is in the SFTP layer, not the TCP layer (which I've investigated by increasing buffers) or the SSH layer (see above).

----

After thinking about it a bit more, I think I see where the problem is.  In connection Read

https://github.com/pkg/sftp/blob/d2aa419835609d317c34baae11bc345d1f9971a7/client.go#L768

The Read function does lots of good windowing receives.  However it only does it for that one buffer.  Each new Read will have to start the windowing process all over again.

Rclone uses buffer sizes of 1MB by default.  Experimenting with that gives this

| rclone buffer size | speed |
| ----------------------- | -------- |
| 64k                       | 320 kB/s |
| 256k                     | 540 kB/s |
| 1 MB                    | 1.1 MB/s |
| 4 MB                    | 1.3 MB/s |
| 16MB                   | 1.5 MB/s |

So the buffer size has an effect but it doesn't get us 19 MB/s like we desire.

What I think needs to happen is that the logic in the Read function gets moved into a background go-routine so it is running continuously, just passing buffers to the Read function as the user desires so it can maintain the window and keep that long pipe full.

Thoughts?

**Comment by ncw** [comment]
PS Can we-reopen this issue please?

**Comment by puellanivis** [comment]
A full-file transfer over SFTP is not just a simple SSH-encapsulated bye stream. And testing its performance relative to a simple SSH-encapsulated cat command is comparing apples to oranges. Basically, you cannot just remove the SFTP protocol overhead and say that should be the ideal target performance that we should shoot for.

Even at best performance, SFTP clients and servers do not typically match the speed of a bare byte stream over SSH, without very creative and clever workarounds built into limitations of the SFTP protocol. In fact, it’s been a rubric that the fastest way to transfer a set of files is often to tar the files up, pipe that to an ssh connection which then tars out the results on the other side.

There are a number of complicated reasons for all of this, but it essentially breaks down to SFTP not being just a raw file transfer protocol like HTTP, FTP, or even a byte stream encapsulated over SSH. Rather, it is actually a network filesystem API implemented on top of SSH. As such, each read of a file needs to be made through a separate request, and must be individually acknowledged in the protocol.

**Comment by eikenb** [comment]
I'll reopen the issue for now, due to the current conversation. If it stops w/o some actionable item I'll re-close.

**Comment by eikenb** [comment]
So the problem is when you are transferring a large number of small files the window size keeps getting reset so it never reaches optimal levels.

Clients usually implement this in a very synchronous/stateful manner where it opens 1 file, tranfers it, then closes it. Just repeating that for all files. To keep the data flowing and the window expanding it seems like it would need to change this to have the client send multiple opens and start sending data for all of them, opening new ones as the old ones close. The protocol supports this with its use of shared client/server file handles.

@ncw Thanks for your work on rclone by the way, I just started using it for offsite/cloud backups and it works great.

**Comment by eikenb** [comment]
It may not be relevant at all, but one thing I wanted to be sure of... you are aware of the issue with using bufio w/ sftp's client.ReadFrom. The issue was raised in #125 and the only resolution for the issue was to document it; https://godoc.org/github.com/pkg/sftp#example-File-ReadFrom-Bufio

Just on the off chance this is another occurrence of that problem.

**Comment by eikenb** [comment]
Sorry for the misunderstanding but I think I understand the problem now. I was thinking of Read as a full-file action, when it in fact works only on part of the file at a time. So It reads until it fills the []byte and then returns, which 0's the window even though there are many more Reads coming.

> Do you want me to have a go at a proof of concept patch?

Yes. Please. Seeing what you mean in code would be a big help. As is I'm not sure how I'd implement what you are thinking  (the background process to keep the window open)... so any help bootstrapping my brain would be great.

Thanks.

**Comment by puellanivis** [comment]
I see what you’re saying about the Read opening a window, transmitting the whole Read operation with a good window, but then blocking all further Reads until the window fills and then the next Read can begin, but naturally has to build another window again. But I don’t see how this could really be solved internally to the sftp package without just assuming that the client wants more data and using Read-Aheads.

So, while having this sliding window might allow one to keep the pipe full more, but it would also background perform pre-reads that might not end up being using. This is common on local filesystems as the bandwidth is typically considered free (after all, it goes unused a lot), but over a network, and especially the internet, that bandwidth has a non-negligible cost.

In many ways, I would see “a better solution” being that if the program wants the entire file, that the program using this package would get the size of the file and then read the whole thing in a single Read operation. But this won’t really works for extremely large files (where file size approaches the size of available physical memory on the machine) and so I suppose Read-Ahead would be a better choice here.

Or perhaps alternatively just include an option that asks for Read-Ahead, or disables Read-Ahead? But then we end up with two complex codepaths, instead of just one. 😕 Of course, there can never be any easy answers.

**Comment by ncw** [comment]
I sat down to have a look at the source code to see how to implement my idea and a few minutes later I realised that this idea has already been implemented in the `WriteTo` function.  The `WriteTo` function keeps the window open while reading from the SFTP file and writing to the client's file.

https://github.com/pkg/sftp/blob/d2aa419835609d317c34baae11bc345d1f9971a7/client.go#L861-L863

So I adjusted rclone to use `WriteTo`.  This involved a bit of messing about with `io.Pipe` to convert the `io.Writer` that you give to `WriteTo` back into an `io.Reader` which I need for the rclone internals.  (I've done this before so it only took a few minutes).

And voila - the speed is now at the 19MB/s I was hoping for over a 100mS link.  This is on a par with the `sftp` binary :-)

Doubtless this can be improved by tweaking buffers etc, but to have it on a par with `sftp` is good enough for the moment.

I'm already using the `ReadFrom` function to upload files with rclone and that works at about 19 MB/s too (I wish I'd noticed this before!).

So the lesson learned is use `WriteTo` and `ReadFrom` if you have links with long latency and wish to stream transfers quickly instead of the `Read` and `Write` methods.

It might be worth mentioning that in the docs - shall I prepare a PR?
